src/main/java/com/microsoft/azure/relay/AsyncSemaphore.java

```diff
@@ -89,6 +91,10 @@ int availablePermits() {
 		});
 	}
 	
+	public <T> CompletableFuture<T> runInsideLockAsync(Duration timeout, AutoShutdownScheduledExecutor executor, Supplier<? extends CompletionStage<T>> supplier) {
```

Comment:
>runInsideLockAsync [](start = 33, length = 18)

I'm not attached to this name. Please make any suggestions you have for better names.

Reply:
A few I could think of:

- lockThenCompose[Async]
- lockThenRun[Async]
- runInsideLock[Async]

---
In reply to: [259040952](https://github.com/Azure/azure-relay-java/pull/20#discussion_r259040952) [](ancestors = 259040952)

Reply:
I feel like we should have "lock", "compose", and  "release"  all somehow jammed together, but that would be a long and awkward name. I'm ok with "lockThenComposeAsync"

---
In reply to: [259043911](https://github.com/Azure/azure-relay-java/pull/20#discussion_r259043911) [](ancestors = 259043911,259040952)

Reply:
The term 'inside' was meant to indicate it acquires, runs whatever code inside, and then releases.

---
In reply to: [259060038](https://github.com/Azure/azure-relay-java/pull/20#discussion_r259060038) [](ancestors = 259060038,259043911,259040952)

Reply:
If you like lockThenComposeAsync I'll use that.

---
In reply to: [259061669](https://github.com/Azure/azure-relay-java/pull/20#discussion_r259061669) [](ancestors = 259061669,259060038,259043911,259040952)

Reply:
What do you think of "synchronizedAsync", since the default synchronized keyword implies lock and release inside the block?

---
In reply to: [259061786](https://github.com/Azure/azure-relay-java/pull/20#discussion_r259061786) [](ancestors = 259061786,259061669,259060038,259043911,259040952)

Reply:
lockThenCompose and lockThenApply have been settled upon.

---
In reply to: [259063410](https://github.com/Azure/azure-relay-java/pull/20#discussion_r259063410) [](ancestors = 259063410,259061786,259061669,259060038,259043911,259040952)